Source/WebCore/html/ValidatedFormListedElement.h

```diff
@@ -0,0 +1,166 @@
+    WEBCORE_EXPORT bool checkValidity(Vector<RefPtr<ValidatedFormListedElement>>* unhandledInvalidControls = nullptr);
+    bool reportValidity();
+    RefPtr<HTMLElement> focusableAnchorElement();
+    void reportNonFocusableControlError();
+    WEBCORE_EXPORT void focusAndShowValidationMessage(Ref<HTMLElement> anchor);
+    bool isShowingValidationMessage() const;
+    WEBCORE_EXPORT bool isFocusingWithValidationMessage() const;
+    // This must be called when a validation constraint or control value is changed.
+    void updateValidity();
+    void setCustomValidity(const String&) override;
+
+    void setAncestorDisabled(bool isDisabled);
+    virtual void reset() { }
+
+    virtual bool supportsReadOnly() const { return false; }
+    bool isDisabled() const { return m_disabled || m_disabledByAncestorFieldset; }
+    bool hasReadOnlyAttribute() const { return m_hasReadOnlyAttribute; }
+
+    // This must be called any time the result of willValidate() has changed.
+    bool isValidFormControlElement() const { return m_isValid; }
+
+    bool isEnumeratable() const override { return false; }
+
+    bool wasInteractedWithSinceLastFormSubmitEvent() const { return m_wasInteractedWithSinceLastFormSubmitEvent; }
+    void setInteractedWithSinceLastFormSubmitEvent(bool);
+
+    bool matchesValidPseudoClass() const;
+    bool matchesInvalidPseudoClass() const;
+    bool matchesUserInvalidPseudoClass() const;
+    bool matchesUserValidPseudoClass() const;
+
+    bool isCandidateForSavingAndRestoringState() const;
+    virtual bool shouldAutocomplete() const;
+    virtual bool shouldSaveAndRestoreFormControlState() const { return false; }
+    virtual FormControlState saveFormControlState() const;
+    virtual void restoreFormControlState(const FormControlState&) { } // Called only if state is not empty.
+    virtual const AtomString& formControlType() const = 0;
+
+protected:
+    virtual bool computeWillValidate() const;
+    void updateWillValidateAndValidity();
+    bool disabledByAncestorFieldset() const { return m_disabledByAncestorFieldset; }
+
+    bool validationMessageShadowTreeContains(const Node&) const;
+
+    virtual void disabledStateChanged();
+    virtual void disabledAttributeChanged();
+    virtual void readOnlyStateChanged();
+
+    void parseAttribute(const QualifiedName&, const AtomString&) override;
+    void insertedIntoAncestor(Node::InsertionType, ContainerNode&);
+    void removedFromAncestor(Node::RemovalType, ContainerNode&);
+    virtual void didFinishInsertingNode();
+    void restoreFormControlStateIfNecessary();
+
+    void willChangeForm() override;
+    void didChangeForm() override;
+
+    void parseDisabledAttribute(const AtomString&);
+    void parseReadonlyAttribute(const AtomString&);
+
+    void setDataListAncestorState(TriState);
+    void syncWithFieldsetAncestors(ContainerNode* insertionNode);
+
+private:
+    bool computeIsDisabledByFieldsetAncestor() const;
+    virtual HTMLElement* anchorElement() = 0;
```

Comment:
It's very misleading to call this anchorElement given we also have HTMLAnchorElement.
Why not validationAnchorElement?

Reply:
Renamed everywhere I've found it, except for `ValidationMessage` where it's just `element`.